python/python/arch.c

```diff
@@ -0,0 +1,398 @@
+	r_return_val_if_fail (as && op, false);
+	// TODO
+//	R_LOG_WARN ("py_arch_encode not implemented");
+	return true;
+}
+
+static bool py_arch_decode(RArchSession *as, RAnalOp *op, RAnalOpMask mask) {
+	r_return_val_if_fail (as && op, false);
+	// WIP
+//	R_LOG_INFO ("py_arch_decode called");
+	bool res = false;
+	if (py_arch_decode_cb) {
+		Py_buffer pybuf = {
+			.buf = (void *) op->bytes,
+			.len = op->size,
+			.readonly = 1,
+			.ndim = 1,
+			.itemsize = 1
+		};
+		PyObject *memview = PyMemoryView_FromBuffer (&pybuf);
+		PyObject *arglist = Py_BuildValue ("(NK)", memview, op->addr);
+		PyObject *result = PyObject_CallFunction (py_arch_decode_cb, "NK", memview, op->addr);
+		if (result) {
+			if (PyList_Check (result)) {
+				PyObject *len = PyList_GetItem (result, 0);
+				PyObject *dict = PyList_GetItem (result, 1);
+				if (len && dict) {
+					if (PyDict_Check (dict)) {
+						op->size = PyNumber_AsSsize_t (len, NULL);
+						op->mnemonic = r_str_new (getS (dict, "mnemonic"));
+						op->cycles = (getI (dict, "cycles"));
+						op->type = (getI (dict, "type"));
+						op->jump = (getI (dict, "jump"));
+						op->fail = (getI (dict, "fail"));
+						op->eob = (getB (dict, "eob"));
+						res = true;
+					}
+				}
+			}
+			Py_DECREF (result);
+		} 
+		Py_DECREF (arglist);
+		Py_DECREF (memview);
+	} else {
+		R_LOG_WARN ("arch plugin does not implement decode");
+	}
+
+	if (PyErr_Occurred ()) {
+		PyErr_Print ();
+	}
+	return res;
+}
+
+static bool py_arch_patch(RArchSession *as, RAnalOp *op, RArchModifyMask mask) {
+	r_return_val_if_fail (as && op, false);
+	// TODO?
+//	R_LOG_WARN ("py_arch_patch not implemented");
+	return true;
+}
+
+static char *py_arch_mnemonics(RArchSession *as, int id, bool json) {
+	r_return_val_if_fail (as, NULL);
+	// TODO?
+//	R_LOG_WARN ("py_arch_mnemonics not implemented");
+	return NULL;
+}
+
+static RList *py_arch_preludes(RArchSession *as) {
+	r_return_val_if_fail (as, NULL);
+	// TODO?
+//	R_LOG_WARN ("py_arch_preludes not implemented");
+	return NULL;
+}
+
+static bool py_arch_esilcb(RArchSession *as, RArchEsilAction action) {
+	r_return_val_if_fail (as, false);
+	// TODO
+//	R_LOG_WARN ("py_arch_esilcb not implemented");
+	return true;
+}
+
+void Radare_plugin_arch_free(RArchPlugin *ap) {
+	if (ap) {
+		free ((char *)ap->meta.name);
+		free ((char *)ap->meta.license);
+		free ((char *)ap->meta.desc);
+		free ((char*)ap->arch);
+		free ((char*)ap->cpus);
+		free (ap);
+	}
+}
+
+RArchPlugin py_arch_plugin = {
+	.init = py_arch_init,
+	.fini = py_arch_fini,
+	.info = py_arch_info,
+	.regs = py_arch_regs,
+	.encode = py_arch_encode,
+	.decode = py_arch_decode,
+	.patch = py_arch_patch,
+	.mnemonics = py_arch_mnemonics,
+	.preludes = py_arch_preludes,
+	.esilcb = py_arch_esilcb
+};
+
+PyObject *Radare_plugin_arch(Radare* self, PyObject *args) {
+	PyObject *arglist = Py_BuildValue ("(i)", 0);
+	PyObject *o = PyObject_CallObject (args, arglist);
+	if (o == NULL) {
```

Comment:
```suggestion
	if (!o) {
```

for consistency with other nullchks